poc/modules/scoring.py

```diff
@@ -0,0 +1,146 @@
+"""High level implementation of scoring modules.
+
+These are pytorch modules that take a list of instances and score each object
+within each instance, taking its context into account. Only the high-level
+assembly is done in this module.
+
+One can easily derive ranking and choice estimators from a scorer, see
+estimators.py for that.
+"""
+
+import torch.nn as nn
+import torch
+
+from modules.embedding import DeepSet
+from modules.embedding import LinearZerothOrderEmbedding
+from modules.feta_support import FirstOrderUtility
+from modules.feta_support import MeanAggregatedUtility
+from modules.utilities import PairwiseLinearUtility
+
+
+class FETAScoring(nn.Module):
+    """Map instances to scores with the FETA approach.
+
+    >>> from modules.utilities import PairwiseMockUtility
+    >>> scoring = FETAScoring(
+    ...     n_features=2,
+    ...     pairwise_utility_module=PairwiseMockUtility,
+    ... )
+
+    Now let's define some problem instances.
+
+    >>> object_a = [0.5, 0.8]
+    >>> object_b = [1.5, 1.8]
+    >>> object_c = [2.5, 2.8]
+    >>> object_d = [3.5, 3.6]
+    >>> object_e = [4.5, 4.6]
+    >>> object_f = [5.5, 5.6]
+    >>> # instance = list of objects to rank
+    >>> instance_a = [object_a, object_b, object_c]
+    >>> instance_b = [object_d, object_e, object_f]
+    >>> import torch
+    >>> instances = torch.tensor([instance_a, instance_b])
+
+    >>> scoring(instances)
+    tensor([[ 4.6000,  6.6000,  8.6000],
+            [16.2000, 18.2000, 20.2000]])
+    """
+
+    # TODO use a more powerful pairwise utility module
+    def __init__(self, n_features, pairwise_utility_module=PairwiseLinearUtility):
+        super().__init__()
+        self.mean_aggregated_utilty = MeanAggregatedUtility(
```

Comment:
In principle we can even think about making this modular. One could use different aggregation functions here or even a learned aggregation operator.

Reply:
Yes, I think that would be interesting :) The composition architecture should make experiments like that much easier.